sms_wsj/database/dynamic_mixing.py

```diff
@@ -0,0 +1,301 @@
+    {'audio_path': {'original_source': ['.../wsj_8k_zeromean/13-16.1/wsj1/si_dt_20/4k2/4k2c031a.wav',
+                                        '.../wsj_8k_zeromean/13-16.1/wsj1/si_dt_20/4k3/4k3c0316.wav'],
+                    'rir': ['.../rirs/cv_dev93/0/h_0.wav',
+                            '.../rirs/cv_dev93/0/h_1.wav']},
+     'dataset': 'cv_dev93_rng1',
+     'example_id': '0_4k2c031a_4k3c0316',
+     'gender': ['female', 'female'],
+     'kaldi_transcription': ['THE S. E. C. IS MANEUVERING TO E- CURB WHAT FUNDS '
+                             'CAN SAY IN NEWSLETTERS JUST AS HOLDERS ARE DEMANDING '
+                             'MORE INFORMATION',
+                             'IN MISSOURI STATE PARTY LEADERS ARE ACTIVELY '
+                             'COURTING DEMOCRATS WHO VOTED FOR MR. REAGAN'],
+     'log_weights': [1.7891449809156743, -1.7891449809156748],
+     'num_samples': {'observation': 59008, 'original_source': [58326, 59008]},
+     'num_speakers': 2,
+     'offset': [659, 0],
+     'room_dimensions': [[8.169], [5.905], [3.073]],
+     'sensor_position': [[4.015, 3.973, 4.03, 4.129, 4.172, 4.115],
+                         [3.265, 3.175, 3.093, 3.102, 3.192, 3.274],
+                         [1.55, 1.556, 1.563, 1.563, 1.558, 1.551]],
+     'sound_decay_time': 0.387,
+     'source_id': ['4k2c031a', '4k3c0316'],
+     'source_position': [[3.312, 3.0], [1.921, 2.379], [1.557, 1.557]],
+     'speaker_id': ['4k2', '4k3']}
+    """
+    def get_sources(ex):
+        num_speakers = len(ex['speaker_id'])
+
+        for spk_idx in range(num_speakers):
+            example_id = ex['source_id'][spk_idx]
+            yield (
+                example_id,
+                {
+                    'audio_path': {
+                        'observation': (ex['audio_path'].get('original_source') or
+                                        ex['audio_path']['speech_source'])[
+                            spk_idx],
+                    },
+                    'example_id': example_id,
+                    **{
+                        k: ex[k][spk_idx]
+                        for k in ['gender', 'kaldi_transcription', 'speaker_id']
+                    },
+                    'num_samples': (ex['num_samples'].get('original_source') or
+                                    ex['num_samples']['speech_source'])[spk_idx],
+                },
+            )
+
+    source_ds = lazy_dataset.new(
+        dict(list(filter_duplicates(ds.map(get_sources).unbatch()))))
+
+    def get_rir_ex(ex):
+        example_id = ex['example_id'].split('_')[0]
+        return (
+            example_id,
+            {
+                'audio_path': {'rir': ex['audio_path']['rir']},
+                'example_id': example_id,
+                **{
+                    k: ex[k]
+                    for k in
+                    ['room_dimensions', 'sound_decay_time', 'source_position',
+                     'sensor_position']
+                },
+            }
+        )
+
```

Comment:
Until yet, there is no case, where a RIR is used multiple times.
At the moment, using a RIR multiple times is problematic, because the uniqueness of the example ID is ensured from the integer that is used to identify a RIR. 

Now, where you mention this, I see, that `filter_duplicates` is not necessary, because the example ID is the key inside the dict and the dict will remove duplicates.